### PR TITLE
Record: 11L Gradient-Guided Adaptive Quant + EMA + Sliding Eval (val_bpb=1.1396)

### PR DESCRIPTION
**Pull request by albertorkive**
## Record: 11L Gradient-Guided Adaptive Quant + EMA + Sliding Eval (val_bpb: 1.1396)

**val_bpb: 1.1396** (sliding window, stride=64, full coverage) | **15.9 MB** | 8xH100 SXM, 600s

### Key Techniques

| Technique | Impact |
|-----------|--------|
| **Gradient-guided adaptive quantization** | Per-tensor int5/int6/int7 bit assignment based on gradient sensitivity. Top 45% get int7, middle 40% get int6, bottom 15% get int5. |
| **EMA from init** (alpha=0.997) | Exponential moving average tracking from initialization, loaded at eval time |
| **Adaptive warmdown** | Warmdown auto-caps at 55% of estimated total steps based on step timing, for robustness across hardware |
| **Sliding window eval** (stride=64) | Full validation set coverage (~121K windows/GPU), scoring only last 64 tokens per window |
| SmearGate | Residual mixing with learnable gate |
| NTK-aware RoPE | Rotary position encoding with NTK interpolation |
| XSA last 4 layers | Cross-sequence attention on final 4 transformer layers |
| Orthogonal init | Orthogonal weight initialization |
| Muon optimizer | momentum=0.99, WD=0.04, grad_clip=0.3 |

### Architecture

- 11 layers, 512 dim, 8 heads, 4 KV heads (GQA)
- MLP 3x expansion (hidden=1536), relu^2 activation
- Tied input/output embeddings
- 26.5M parameters

### Results

| Metric | Value |
|--------|-------|
| Pre-quant val_bpb | 1.1339 |
| **Post-quant sliding val_bpb (s64)** | **1.1396** |
| Steps completed (600s cap) | 5,347 |
| Step time | 112ms |
| Model params | 26,502,232 |
| Artifact size | 15,913,419 bytes |
| Quant distribution | {int5: 12, int6: 31, int7: 36} |

### Run Command

```bash
pip install sentencepiece zstandard
python3 data/cached_challenge_fineweb.py
torchrun --nproc_per_node=8 records/track_10min_16mb/2026-03-22_11L_GradQuant_EMA_SlidingEval/train_gpt.py
```

All hyperparameters baked into train_gpt.py. No env vars needed.

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
# Sibling Draft Review for PR #422

**Date:** 2026-04-12  
**Reviewer:** Claude Agent (for Mato/@MatoTeziTanka)  
**NEEDS_LLM_REVIEW:** Yes  
**Status:** Ready for detailed code review

## PR Summary

**Title:** Record: 11L Gradient-Guided Adaptive Quant + EMA + Sliding Eval (val_bpb=1.1396)  
**State:** OPEN  
**Mato blocking comment:** NO  

## Marker Analysis

| Marker | Found | Notes |
|--------|-------|-------|
| target_in_key | False | Custom loss key injection pattern |
| TTT | False | Test-Time Training integration |
| SLOT | False | Slot-based attention variant |
| custom_tokenizer | True | 3 patterns detected |

## Architecture Changes

**train_gpt.py modifications:** ~1368 lines changed  
**Quantization:** True  
**EMA:** True  
**LoRA:** False  
**GPTQ:** False

## Assessment

⚠ REQUIRES_REVIEW - Custom modifications detected:
- Custom tokenizer patterns found (3 instances)

## Recommendation

**REVIEW** — Detailed code inspection needed before merge

## Review Checklist

- [ ] Code style and formatting (PEP 8, consistency)
- [ ] Numerical stability of modifications
- [ ] Integration with existing pipeline
- [ ] Reproducibility (seed handling, RNG control)
- [ ] Performance metrics reported correctly
- [ ] No undocumented dependencies

## Next Steps

1. Full train_gpt.py code review
2. Validate loss computation and gradients
3. Check for integration issues with existing pipeline
4. Verify metrics reproducibility

---
**Generated for:** Mato (@MatoTeziTanka)  
**Scope:** NEEDS_LLM_REVIEW sweep  
**Action:** Draft only—no posts to GitHub


**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** MERGE-pending standard record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Classification via sibling-session agent (Haiku-backed). This review was drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*

**Comment by MatoTeziTanka** [comment]
# Sibling Draft Review for PR #422

**Date:** 2026-04-12  
**Reviewer:** Claude Agent (for Mato/@MatoTeziTanka)  
**NEEDS_LLM_REVIEW:** Yes  
**Status:** Ready for detailed code review

## PR Summary

**Title:** Record: 11L Gradient-Guided Adaptive Quant + EMA + Sliding Eval (val_bpb=1.1396)  
**State:** OPEN  
**Mato blocking comment:** NO  

## Marker Analysis

| Marker | Found | Notes |
|--------|-------|-------|
| target_in_key | False | Custom loss key injection pattern |
| TTT | False | Test-Time Training integration |
| SLOT | False | Slot-based attention variant |
| custom_tokenizer | True | 3 patterns detected |

## Architecture Changes

**train_gpt.py modifications:** ~1368 lines changed  
**Quantization:** True  
**EMA:** True  
**LoRA:** False  
**GPTQ:** False

## Assessment

⚠ REQUIRES_REVIEW - Custom modifications detected:
- Custom tokenizer patterns found (3 instances)

## Recommendation

**REVIEW** — Detailed code inspection needed before merge

## Review Checklist

- [ ] Code style and formatting (PEP 8, consistency)
- [ ] Numerical stability of modifications
- [ ] Integration with existing pipeline
- [ ] Reproducibility (seed handling, RNG control)
- [ ] Performance metrics reported correctly
- [ ] No undocumented dependencies

## Next Steps

1. Full train_gpt.py code review
2. Validate loss computation and gradients
3. Check for integration issues with existing pipeline
4. Verify metrics reproducibility

---
**Generated for:** Mato (@MatoTeziTanka)  
**Scope:** NEEDS_LLM_REVIEW sweep  
**Action:** Draft only—no posts to GitHub


**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** MERGE-pending standard record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Classification via sibling-session agent (Haiku-backed). This review was drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*

**Comment by MatoTeziTanka** [comment]
### Community Review — Record: 11L Gradient-Guided Adaptive Quant + EMA + Sliding Eval (val_bpb=1.1396)

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache


Head SHA: 2afa1312eb2b892162e0e8899d826e8fa661ff91

---

## Analysis

### 1. N-gram family bug (CLOSE trigger)

**CLEAN.** The submission uses a `BigramHash`-exempt standard eval pipeline. Searched `train_gpt.py` exhaustively for `ngram`, `n_gram`, `bigram`, `trigram`, `hash`, and all return zero hits. The byte-level BPB computation in `eval_val` uses `base_bytes_lut`, `has_leading_space_lut`, and `is_boundary_token_lut` — these are standard sentencepiece byte-count look-up tables, not n-gram predictors. No hash lookup on target tokens for prediction purposes exists anywhere in the file. No CLOSE trigger.

### 2. Pre-Quant TTT (CLOSE trigger)

**CLEAN.** There is no TTT (test-time training) pass anywhere in the file. The string `ttt`, `test_time`, `finetune`, and `adamw` applied to `val_tokens` all return zero hits. The only optimizer usage on validation data is the read-only `eval_val()` function which runs entirely under `torch.inference_mode()` with `.detach()` on outputs. After training ends, quantized weights are loaded and `eval_val` is called once more for the roundtrip check — still inference-only, no parameter updates. No pre-quant TTT on val data. No CLOSE trigger.

### 3. Legal TTT (CLEAN check per PR #1413)

**Not present, therefore CLEAN.** No TTT pass exists at all — neither legal nor illegal. The reference pattern (score-first per chunk under `inference_mode()` before any parameter update) is moot because no eval-time adaptation is performed.

### 4. Scored-region SLOT (HOLD trigger)

**Not present.** No SLOT (standard or causal), no scored-region delta optimization. No HOLD trigger.

### 5. Pure neural (CLEAN)

**CLEAN.** The model is a standard autoregressive transformer (11L, 512d, GQA 8H/4KV, tied embeddings, RoPE, logit softcap). Techniques include EMA, SmearGate residual mixing, XSA on last 4 layers, gradient-guided adaptive quantization, and NTK-aware RoPE — all pure neural, no external lookup tables or cache-based inference augmentation.

---

## Summary

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source. If this review misread your code, please call it out so I can re-audit manually.*
